src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -213,36 +216,60 @@ internal static void EndJsonPaddingIfRequired(IJsonWriter jsonWriter, ODataMessa
         /// <param name="innerErrorPropertyName">The property name for the inner error property.</param>
         /// <param name="recursionDepth">The number of times this method has been called recursively.</param>
         /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
-        private static void WriteInnerError(IJsonWriter jsonWriter, ODataInnerError innerError, string innerErrorPropertyName, int recursionDepth, int maxInnerErrorDepth)
+        /// <param name="skipNullProperties">if set, stacktrace, type, and message will not be written if they are empty strings on ODataInnerError.</param>
+
+        private static void WriteInnerError(IJsonWriter jsonWriter, ODataInnerError innerError, string innerErrorPropertyName, int recursionDepth, int maxInnerErrorDepth, bool skipNullProperties)
         {
             Debug.Assert(jsonWriter != null, "jsonWriter != null");
             Debug.Assert(innerErrorPropertyName != null, "innerErrorPropertyName != null");
 
             ValidationUtils.IncreaseAndValidateRecursionDepth(ref recursionDepth, maxInnerErrorDepth);
 
-            // "innererror": {
+            // "innererror":
             jsonWriter.WriteName(innerErrorPropertyName);
             jsonWriter.StartObjectScope();
 
-            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
-            ////       to stay compatible with Astoria.
+            if (innerError.Properties != null)
+            {
+                foreach (KeyValuePair<string, ODataValue> pair in innerError.Properties)
+                {
+                    jsonWriter.WriteName(pair.Key);
+                    jsonWriter.WriteODataValue(pair.Value);
+                }
+            }
 
-            // "message": "<message>"
-            jsonWriter.WriteName(JsonConstants.ODataErrorInnerErrorMessageName);
-            jsonWriter.WriteValue(innerError.Message ?? string.Empty);
+            if (!skipNullProperties)

```

Comment:
>f (!skipNullProperties [](start = 13, length = 22)

If Properties has null values, the "SkipNullProperties" setting can't effect them?